cpp/ql/lib/semmle/code/cpp/rangeanalysis/SimpleRangeAnalysis.qll

```diff
@@ -1530,6 +1530,29 @@ private predicate isUnsupportedGuardPhi(Variable v, RangeSsaDefinition phi, Vari
   )
 }
 
+/**
+ * Gets the upper bound of the expression, if the expression is guarded.
+ * An upper bound can only be found, if a guard phi node can be found, and the
+ * expression has only one immediate predecessor.
+ */
+private float getGuardedUpperBound(VariableAccess guardedAccess) {
+  exists(
+    RangeSsaDefinition def, StackVariable v, VariableAccess guardVa, Expr guard, boolean branch
+  |
+    def.isGuardPhi(v, guardVa, guard, branch) and
+    // If the basic block for the variable access being examined has
+    // more than one predecessor, the guard phi node could originate
+    // from one of the predecessors. This is because the guard phi
+    // node is attached to the block at the end of the edge and not on
+    // the actual edge. It is therefore possible to determine which
+    // edge the guard phi node belongs to. The predicate below ensures
+    // that there is one predecessor, albeit somewhat conservative.
+    exists(unique(BasicBlock b | b = def.(BasicBlock).getAPredecessor())) and
```

Comment:
I think this can be strengthened a bit to
```ql
    exists(BasicBlock pred |
      pred = guard.getBasicBlock() and
      forall(BasicBlock bb | bb = def.(BasicBlock).getAPredecessor() | bbDominates(pred, bb))
    )
```

that should also allow for cases such as
```cpp
if (x < 0) {
  do {
    out(x); // has two predecessors
  } while ...
}
```

(see also https://github.com/github/codeql/blob/main/csharp/ql/lib/semmle/code/csharp/controlflow/BasicBlocks.qll#L458).

Reply:
I read this as: All predecessors to the `def` block must dominate the guard expression. 
Then when examining the variable access for the last use of x, in the following example:
```c
if (x > 10) {  // guard
  use(x); // B
}
use(x) // def
```
We have 
`def` has two predecessors: `B`, `guard`. 
B is dominated by `guard`.
`guard` is dominated by `guard`.

So we find that the guard phi node `def` is a usable candidate, and hence x must be less equal to 10

Reply:
Sorry, I got it wrong. It should have been
```ql
    forall(BasicBlock pred |
      pred = def.(BasicBlock).getAPredecessor() and pred != guard.getBasicBlock()
    |
      bbDominates(def, pred)
    )
```

Reply:
It may not make a difference in practice (right now) because in the CFG for the example
```cpp
if (x < 0) {
  do {
    out(x);
  } while (b)
}
```
the `do` node itself is included in the CFG as the successor of `x < 0`, and the return edge from `b` does not go to `do`, but instead directly to the body of the loop. However, I would still prefer the change, as it will also work if someone decided to remove the `do` node.

Reply:
Makes sense. I will add the change.

Reply:
@jbj found an example where the guard phi has two predecessors which are both in the dominance domain of the guard phi:

```c
void test(int x) {
  if (x >= 10) {
    return;
  }
  // The basic below has two predecessors.
label:
  out(x);
  goto label;
}
```
I've added the test case to the PR